consensus/istanbul/core/core.go

```diff
@@ -344,10 +381,46 @@ func (c *core) waitForDesiredRound(r *big.Int) {
 
 func (c *core) updateRoundState(view *istanbul.View, validatorSet istanbul.ValidatorSet, roundChange bool) {
 	// TODO(Joshua): Include desired round here.
-	if roundChange && c.current != nil {
-		c.current = newRoundState(view, validatorSet, nil, c.current.pendingRequest, c.current.preparedCertificate, c.backend.HasBadProposal)
+	if c.current != nil {
+		var commits *messageSet
+		if roundChange {
+			commits = c.current.ParentCommits
+			// pass over the current certificate in round changes
+			c.current = newRoundState(view, validatorSet, nil, c.current.pendingRequest, c.current.preparedCertificate, commits, c.backend.HasBadProposal)
+		} else {
+			// if it was not a round change (ie. a sequence change),
+			// we modify our ParentCommits to the last sequence's commits
+			lastProposal, _ := c.backend.LastProposal()
+			// TODO(gakonst): What are the exact cases that the current round's preprepare will be nil?
+			if c.current.Preprepare != nil {
+				preprepareProposal := c.current.Preprepare.Proposal
+				if lastProposal.Hash() == preprepareProposal.Hash() {
+					c.logger.Debug("hashes-matched", "num", preprepareProposal.Number(), "hash", preprepareProposal.Hash().String())
+					// if it was not a round change (ie. a sequence change) and the
+					// preprepare proposal hash matches the chain head,
+					// we use this sequence's commits as the ParentCommits field in the next round
+					commits = c.current.Commits
+				} else {
+					c.logger.Debug("preprepare", "num", preprepareProposal.Number(), "hash", preprepareProposal.Hash().String())
+					c.logger.Debug("lastproposal", "num", lastProposal.Number(), "hash", lastProposal.Hash().String())
+					// if the hashes did not match, this means there was a partition and
+					// the lastProposal is different from the one we were listening to.
+					// as a result, we discard ths current round's commits, and initialize an empty
+					// ParentCommits message set, which we use to listen for messages which match the
+					// lastProposal as a parent.
+					commits = newMessageSet(c.backend.Validators(lastProposal))
```

Comment:
Is there an off-by-one error here? i.e. does `Validators` return the validator set that signs off on `lastProposal` or `lastProposal+1`. This is one of the most unnecessarily confusing parts of this code base.